core/templates/dev/head/exploration_editor/ExplorationEditor.js

```diff
@@ -620,49 +624,140 @@ oppia.controller('ExplorationSaveAndPublishButtons', [
+          !explorationCategoryService.savedMemento ||
+          !explorationLanguageCodeService.savedMemento) {
+         $modal.open({
+           templateUrl: 'modals/addExplorationMetadata',
+           backdrop: true,
+           controller: [
+             '$scope', '$modalInstance', 'explorationObjectiveService',
+             'explorationTitleService', 'explorationCategoryService',
+             'explorationStatesService', 'CATEGORY_LIST',
+             'explorationLanguageCodeService',
+             function($scope, $modalInstance, explorationObjectiveService,
+             explorationTitleService, explorationCategoryService,
+             explorationStatesService, CATEGORY_LIST,
+             explorationLanguageCodeService) {
+               $scope.explorationTitleService =
+               explorationTitleService;
+               $scope.explorationObjectiveService =
+               explorationObjectiveService;
+               $scope.explorationCategoryService =
+               explorationCategoryService;
+               $scope.explorationLanguageCodeService =
+               explorationLanguageCodeService;
+
+               $scope.requireTitleToBeSpecified = (
+                 !explorationTitleService.savedMemento);
+               $scope.requireObjectiveToBeSpecified = (
+                 !explorationObjectiveService.savedMemento);
+               $scope.requireCategoryToBeSpecified = (
+                 !explorationCategoryService.savedMemento);
+               $scope.requireLanguageToBeSpecified = (
+                 !explorationLanguageCodeService.savedMemento);
+
+               $scope.metadataList = [];
+
+               $scope.CATEGORY_LIST_FOR_SELECT2 = [];
+
+               for (var i = 0; i < CATEGORY_LIST.length; i++) {
+                 $scope.CATEGORY_LIST_FOR_SELECT2.push({
+                   id: CATEGORY_LIST[i],
+                   text: CATEGORY_LIST[i]
+                 });
+               }
+
+               var _states = explorationStatesService.getStates();
+               if (_states) {
+                 var categoryIsInSelect2 = $scope.CATEGORY_LIST_FOR_SELECT2
+                 .some(
+                   function(categoryItem) {
+                     return categoryItem.id ===
+                     explorationCategoryService.savedMemento;
+                   }
+                 );
+
+                 if (!categoryIsInSelect2) {
+                   $scope.CATEGORY_LIST_FOR_SELECT2.push({
+                     id: explorationCategoryService.savedMemento,
+                     text: explorationCategoryService.savedMemento
+                   });
+                 }
+               }
+
+               $scope.save = function() {
+                 // If no objective has been specified yet, require the creator
+                 // to specify it.
+                 if ($scope.requireObjectiveToBeSpecified) {
+                   if (!explorationObjectiveService.displayed) {
```

Comment:
No, this is still incorrect. Let's say the objective and title are specified but the category isn't. The user clicks "save". The metadataList then becomes ['objective', 'title'] and then an error is thrown for the category. The user fixes the category and clicks save again. The metadataList now becomes ['objective', 'title', 'objective', 'title', 'category'], which isn't what you want.

What I am suggesting is, don't change the metadataList or save anything until you are sure that all the conditions for saving are satisfied. I.e., at the start of the function, make sure that all necessary displayed values exist. Only then, do all the saveDisplayedValue() and metadataList.push() calls.